File: 2000/S5.cpp
#include <bits/stdc++.h>

int main() {
    int n;
    std::cin >> n;

    std::cout << std::fixed << std::setprecision(28);

    std::vector<std::pair<double, double>> sheep;
    sheep.reserve(n);

    for (int i = 0; i < n; ++i) {
        double x, y;
        std::cin >> x >> y;
        sheep.emplace_back(x, y);
    }

    for (int i = 0; i < n; ++i) {
        std::pair<double, double> interval = {0, 1000};

        for (int j = 0; j < n; ++j) {
            if (i == j) {
                continue;
            }

            auto [x1, y1] = sheep[i];
            auto [x2, y2] = sheep[j];

            if (x1 == x2) {
                if (y1 > y2) {
                    interval = {1, 0};
                    break;
                }
                continue;
            } else if (x1 > x2) {
                double min_x = (x2 * x2) / (2 * (-x1 + x2)) + (y2 * y2) / (2 * (-x1 + x2)) - (x1 * x1) / (2 * (-x1 + x2)) - (y1 * y1) / (2 * (-x1 + x2));
                interval.first = std::max(interval.first, min_x);
            } else {
                double max_x = (x2 * x2) / (2 * (-x1 + x2)) + (y2 * y2) / (2 * (-x1 + x2)) - (x1 * x1) / (2 * (-x1 + x2)) - (y1 * y1) / (2 * (-x1 + x2));
                interval.second = std::min(interval.second, max_x);
            }

            if (interval.first > interval.second) {
                break;
            }
        }

        if (interval.first <= interval.second) {
            std::cout << std::fixed << std::setprecision(2);
            std::cout << "The sheep at (" << sheep[i].first << ", " << sheep[i].second << ") might be eaten.\n";
        }
    }

    return 0;
}

File: 2000/S5.py
from decimal import *
getcontext().prec = 28

n = int(input())

sheep = []
for i in range(n):
    x = Decimal(input())
    y = Decimal(input())
    sheep.append((x, y))

for i in range(n):
    interval = [0, 1000]

    for j in range(n):
        if i == j:
            continue
    
        x1, y1 = sheep[i]
        x2, y2 = sheep[j]

        if x1 == x2:
            if y1 > y2:
                interval = [1, 0]
                break
            continue
        elif x1 > x2:
            min_x = x2 ** 2 / (2 * (-x1 + x2)) + y2 ** 2 / (2 * (-x1 + x2)) - x1 ** 2 / (2 * (-x1 + x2)) - y1 ** 2 / (2 * (-x1 + x2))
            interval[0] = max(interval[0], min_x)
        else:
            max_x = x2 ** 2 / (2 * (-x1 + x2)) + y2 ** 2 / (2 * (-x1 + x2)) - x1 ** 2 / (2 * (-x1 + x2)) - y1 ** 2 / (2 * (-x1 + x2))
            interval[1] = min(interval[1], max_x)
        
        if interval[0] > interval[1]:
            break
    
    if interval[0] <= interval[1]:
        print("The sheep at (" + '{:.2f}'.format(sheep[i][0]) + ", " + '{:.2f}'.format(sheep[i][1]) + ") might be eaten.")